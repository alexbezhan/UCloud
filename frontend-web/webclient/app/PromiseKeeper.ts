export default class PromiseKeeper {
    public canceledKeeper: boolean = false;
    private promises: Array<CancelablePromise<any>>;

    constructor() {
        this.promises = [];
    }

    public makeCancelable<T>(promise: Promise<T>): CancelablePromise<T> {
        const cancelablePromise: CancelablePromise<T> = {
            hasCanceled_: false,
            isComplete: false,
            promise: new Promise((resolve: any, reject: any) => {
                promise.then(
                    val => {
                        cancelablePromise.isComplete = true;
                        this.cleanup();
                        cancelablePromise.hasCanceled_ ? reject({isCanceled: true}) : resolve(val);
                    },
                    error => {
                        cancelablePromise.isComplete = true;
                        this.cleanup();
                        cancelablePromise.hasCanceled_ ? reject({isCanceled: true}) : reject(error);
                    }
                );
            }),
            cancel: () => cancelablePromise.hasCanceled_ = true,
        };
        this.promises.push(cancelablePromise);
        return cancelablePromise;
    }
    /**
     *  Cancels all promises stored in the promise keeper.
     *  The held promises are cleared from the keeper as they are cancelled and no longer have any function
     */
    public cancelPromises = (): void => {
        this.canceledKeeper = true;
        this.promises.forEach((it) => it.cancel());
        this.promises = [];
    }

    private cleanup = (): void => {
        this.promises = this.promises.filter(it => !it.isComplete);
    }
}

interface CancelablePromise<T> {
    isComplete: boolean;
    promise: Promise<T>;
    cancel: VoidFunction;
    hasCanceled_: boolean;
}
