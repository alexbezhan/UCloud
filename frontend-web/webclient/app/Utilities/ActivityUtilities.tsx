import {ActivityFilter} from "Activity";
import {ScrollRequest} from "Scroll";
import {buildQueryString} from "./URIUtilities";

interface ParamsType extends Omit<ActivityFilter, "maxTimestamp" | "minTimestamp">, Partial<ScrollRequest<number>> {
    maxTimestamp?: number;
    minTimestamp?: number;
}

export const activityQuery = (scroll: ScrollRequest<number>, filter?: ActivityFilter) => {
    const params: ParamsType = {};
    if (!!scroll.offset) params.offset = scroll.offset;
    params.scrollSize = scroll.scrollSize;

    if (filter !== undefined) {
        if (filter.type !== undefined) params.type = filter.type;
        if (filter.collapseAt !== undefined) params.collapseAt = filter.collapseAt;
        if (filter.maxTimestamp !== undefined) params.maxTimestamp = filter.maxTimestamp.getTime();
        if (filter.minTimestamp !== undefined) params.minTimestamp = filter.minTimestamp.getTime();
        if (filter.user !== undefined) params.user = filter.user;
    }

    return buildQueryString("/activity/browse", params);
};

export const activityStreamByPath = (path: string) =>
    `/activity/by-path?path=${encodeURIComponent(path)}`;
