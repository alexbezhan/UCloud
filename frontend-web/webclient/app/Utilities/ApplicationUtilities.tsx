import {
    ApplicationInvocationDescription,
    ApplicationMetadata,
    ApplicationParameter,
    AppState,
    ParameterTypes,
    RunsSortBy
} from "Applications";
import Cloud from "Authentication/lib";
import {SortOrder} from "Files";
import * as React from "react";
import {snackbarStore} from "Snackbar/SnackbarStore";
import {Page} from "Types";
import {addStandardDialog} from "UtilityComponents";
import {errorMessageOrDefault, removeTrailingSlash} from "UtilityFunctions";
import {expandHomeFolder} from "./FileUtilities";

export const hpcJobQueryPost = "/hpc/jobs";

export const hpcJobQuery = (
    id: string,
    stdoutLine: number,
    stderrLine: number,
    stdoutMaxLines: number = 1000,
    stderrMaxLines: number = 1000
) =>
    `/hpc/jobs/follow/${encodeURIComponent(id)}?stdoutLineStart=${stdoutLine}&stdoutMaxLines=${stdoutMaxLines}&stderrLineStart=${stderrLine}&stderrMaxLines=${stderrMaxLines}`;

export function hpcJobsQuery(
    itemsPerPage: number,
    page: number,
    sortOrder?: SortOrder,
    sortBy?: RunsSortBy,
    minTimestamp?: number,
    maxTimestamp?: number,
    filter?: AppState
): string {
    let query = `/hpc/jobs/?itemsPerPage=${itemsPerPage}&page=${page}`;
    if (sortOrder) query = query.concat(`&order=${sortOrder}`);
    if (sortBy) query = query.concat(`&sortBy=${sortBy}`);
    if (minTimestamp != null) query = query.concat(`&minTimestamp=${minTimestamp}`);
    if (maxTimestamp != null) query = query.concat(`&maxTimestamp=${maxTimestamp}`);
    if (filter != null) query = query.concat(`&filter=${filter}`);
    return query;
}

export const hpcFavoriteApp = (name: string, version: string) =>
    `/hpc/apps/favorites/${encodeURIComponent(name)}/${encodeURIComponent(version)}`;

export const hpcFavorites = (itemsPerPage: number, pageNumber: number) =>
    `/hpc/apps/favorites?itemsPerPage=${itemsPerPage}&page=${pageNumber}`;

export const hpcApplicationsQuery = (page: number, itemsPerPage: number) =>
    `/hpc/apps?page=${page}&itemsPerPage=${itemsPerPage}`;

interface HPCApplicationsSearchQuery {
    query: string,
    page: number,
    itemsPerPage: number
}

export const hpcApplicationsSearchQuery = ({query, page, itemsPerPage}: HPCApplicationsSearchQuery): string =>
    `/hpc/apps/search?query=${encodeURIComponent(query)}&page=${page}&itemsPerPage=${itemsPerPage}`;

export const hpcApplicationsTagSearchQuery = ({query, page, itemsPerPage}: HPCApplicationsSearchQuery): string =>
    `/hpc/apps/searchTags?query=${encodeURIComponent(query)}&page=${page}&itemsPerPage=${itemsPerPage}`;

export const cancelJobQuery = `hpc/jobs`;

export const cancelJobDialog = ({jobId, onConfirm, jobCount = 1}: { jobCount?: number, jobId: string, onConfirm: () => void }): void =>
    addStandardDialog({
        title: `Cancel job${jobCount > 1 ? "s" : ""}?`,
        message: jobCount === 1 ? `Cancel job: ${jobId}?` : "Cancel jobs?",
        cancelText: "No",
        confirmText: `Cancel job${jobCount > 1 ? "s" : ""}`,
        onConfirm
    });

export const cancelJob = (cloud: Cloud, jobId: string): Promise<{ request: XMLHttpRequest, response: void }> =>
    cloud.delete(cancelJobQuery, {jobId});

interface FavoriteApplicationFromPage<T> {
    name: string
    version: string
    page: Page<{ metadata: ApplicationMetadata, favorite: boolean } & T>
    cloud: Cloud
}

/**
 * Favorites an application.
 * @param {Application} Application the application to be favorited
 * @param {Cloud} cloud The cloud instance for requests
 */
export async function favoriteApplicationFromPage<T>({name, version, page, cloud}: FavoriteApplicationFromPage<T>): Promise<Page<T>> {
    const a = page.items.find(it => it.metadata.name === name && it.metadata.version === version)!;
    try {
        await cloud.post(hpcFavoriteApp(name, version));
        a.favorite = !a.favorite;
    } catch (e) {
        snackbarStore.addFailure(errorMessageOrDefault(e, `An error ocurred favoriting ${name}`));
    }
    return page;
}


interface StringMap {
    [k: string]: string;
}

interface AllowedParameterKey {
    name: string;
    type: ParameterTypes;
}

interface ExtractParameters {
    nameToValue: StringMap;
    allowedParameterKeys: AllowedParameterKey[];
    siteVersion: number;
}

export const findKnownParameterValues = ({nameToValue, allowedParameterKeys, siteVersion}: ExtractParameters): StringMap => {
    const extractedParameters = {};
    if (siteVersion === 1) {
        allowedParameterKeys.forEach(({name, type}) => {
            if (nameToValue[name] !== undefined) {
                if (typeMatchesValue(type, nameToValue[name])) {
                    extractedParameters[name] = nameToValue[name];
                }
            }
        });
    }
    return extractedParameters;
};

export const isFileOrDirectoryParam = ({type}: { type: string }) => type === "input_file" || type === "input_directory";

const typeMatchesValue = (type: ParameterTypes, parameter: string): boolean => {
    switch (type) {
        case ParameterTypes.Boolean:
            return parameter === "Yes" || parameter === "No" || parameter === "";
        case ParameterTypes.Integer:
            return parseInt(parameter, 10) % 1 === 0;
        case ParameterTypes.FloatingPoint:
            return typeof parseFloat(parameter) === "number";
        case ParameterTypes.Text:
        case ParameterTypes.InputDirectory:
        case ParameterTypes.InputFile:
        case ParameterTypes.SharedFileSystem:
        case ParameterTypes.Peer:
            return typeof parameter === "string";

    }
};

interface ExtractedParameters {
    [key: string]: string | number | boolean |
        { source: string, destination: string; } |
        { fileSystemId: string } |
        { jobId: string }
}

export type ParameterValues = Map<string, React.RefObject<HTMLInputElement | HTMLSelectElement>>;

interface ExtractParametersFromMap {
    map: ParameterValues;
    appParameters: ApplicationParameter[];
    cloud: Cloud;
}

export function extractValuesFromWidgets({map, appParameters, cloud}: ExtractParametersFromMap): ExtractedParameters {
    const extracted: ExtractedParameters = {};
    map.forEach(({current}, key) => {
        const parameter = appParameters.find(it => it.name === key);
        console.log(parameter, current);
        if (!current) return;
        if (!current.value || !current.checkValidity()) return;
        if (!parameter) return;
        switch (parameter.type) {
            case ParameterTypes.InputDirectory:
            case ParameterTypes.InputFile:
                const expandedValue = expandHomeFolder(current.value, cloud.homeFolder);
                extracted[key] = {
                    source: expandedValue,
                    destination: removeTrailingSlash(expandedValue).split("/").pop()!
                };
                return;
            case ParameterTypes.Boolean:
                switch (current.value) {
                    case "Yes":
                        extracted[key] = true;
                        return;
                    case "No":
                        extracted[key] = false;
                        return;
                    default:
                        return;
                }
            case ParameterTypes.Integer:
                extracted[key] = parseInt(current.value, 10);
                return;
            case ParameterTypes.FloatingPoint:
                extracted[key] = parseFloat(current.value);
                return;
            case ParameterTypes.Text:
                extracted[key] = current.value;
                return;
            case ParameterTypes.SharedFileSystem:
                console.log(current.value);
                extracted[key] = {fileSystemId: current.value};
                return;
            case ParameterTypes.Peer:
                extracted[key] = {jobId: current.value};
                return;
        }
    });
    return extracted;
}

export const inCancelableState = (state: AppState) =>
    state === AppState.VALIDATED ||
    state === AppState.PREPARED ||
    state === AppState.SCHEDULED ||
    state === AppState.RUNNING;


export function validateOptionalFields(
    invocation: ApplicationInvocationDescription,
    parameters: ParameterValues
): boolean {
    const optionalErrors = [] as string[];
    const optionalParams = invocation.parameters.filter(it => it.optional && it.visible).map(it =>
        ({name: it.name, title: it.title})
    );
    optionalParams.forEach(it => {
        const param = parameters.get(it.name)!;
        if (!param.current!.checkValidity()) optionalErrors.push(it.title);
    });

    if (optionalErrors.length > 0) {
        snackbarStore.addFailure(
            `Invalid values for ${optionalErrors.slice(0, 3).join(", ")}
                    ${optionalErrors.length > 3 ? `and ${optionalErrors.length - 3} others` : ""}`,
            5000
        );
        return false;
    }

    return true;
}

export function checkForMissingParameters(
    parameters: ExtractedParameters,
    invocation: ApplicationInvocationDescription
): boolean {
    const requiredParams = invocation.parameters.filter(it => !it.optional);
    const missingParameters: string[] = [];
    requiredParams.forEach(rParam => {
        const parameterValue = parameters[rParam.name];
        // Number, string, boolean 
        if (!parameterValue) {
            missingParameters.push(rParam.title);
        } else if (rParam.type === ParameterTypes.InputDirectory || rParam.type === ParameterTypes.InputFile) {
            if (!parameterValue["source"]) {
                missingParameters.push(rParam.title);
            }
        } else if (rParam.type === ParameterTypes.SharedFileSystem) {
            if (!parameterValue["fileSystemId"]) {
                missingParameters.push(rParam.title);
            }
        }
    });

    // Check missing values for required input fields.
    if (missingParameters.length > 0) {
        snackbarStore.addFailure(
            `Missing values for ${missingParameters.slice(0, 3).join(", ")} 
                ${missingParameters.length > 3 ? `and ${missingParameters.length - 3} others.` : ``}`,
            5000
        );
        return false;
    }
    return true;
}
