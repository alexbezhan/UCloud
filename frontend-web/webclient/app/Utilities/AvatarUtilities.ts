export const saveAvatarQuery = "/avatar/update";
export const findAvatarQuery = "/avatar/find";
