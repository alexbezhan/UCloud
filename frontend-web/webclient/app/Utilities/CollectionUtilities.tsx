export function removeEntry<T>(arr: T[], index: number): T[] {
    return arr.slice(0, index).concat(arr.slice(index + 1));
}

export function setsDiffer<T>(s1: Set<T>, s2: Set<T>): boolean {
    if (s1.size !== s2.size) return true;
    let differ = false;
    s1.forEach(it => {
        if (!s2.has(it)) differ = true;
    });
    return differ;
}

type PrimitiveDataTypes = string | number | boolean;

/**
 * @param set
 * @param entry
 */
export function addEntryIfNotPresent(set: Set<PrimitiveDataTypes>, entry: PrimitiveDataTypes): boolean {
    const size = set.size;
    set.add(entry);
    return size !== set.size;
}

export function groupBy<T>(items: T[], keySelector: (t: T) => string): Record<string, T[]> {
    const result: Record<string, T[]> = {};
    items.forEach(item => {
        const key = keySelector(item);
        const newList = result[key] ?? [];
        newList.push(item);
        result[key] = newList;
    });
    return result;
}

export function associateBy<T>(items: T[], keySelector: (t: T) => string): Record<string, T> {
    const result: Record<string, T> = {};
    items.forEach(item => {
        const key = keySelector(item);
        result[key] = item;
    });
    return result;
}

export function takeLast<T>(items: T[], numberOfItems: number): T[] {
    return items.slice(Math.max(0, items.length - numberOfItems));
}
