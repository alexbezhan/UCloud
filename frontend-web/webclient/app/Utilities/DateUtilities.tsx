import {format} from "date-fns/esm";

// Could potentially cause issues with time if user is outside CEST
export const dateToString = (date: number) => format(date, "HH:mm dd/MM/yyyy");
