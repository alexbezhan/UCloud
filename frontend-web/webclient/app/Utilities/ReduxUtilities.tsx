import activity from "Activity/Redux/ActivityReducer";
import * as AppRedux from "Applications/Redux";
import analyses from "Applications/Redux/AnalysesReducer";
import applications from "Applications/Redux/BrowseReducer";
import detailedApplicationSearch from "Applications/Redux/DetailedApplicationSearchReducer";
import * as TaskRedux from "BackgroundTasks/redux";
import dashboard from "Dashboard/Redux/DashboardReducer";
import {initObject} from "DefaultObjects";
import detailedFileSearch from "Files/Redux/DetailedFileSearchReducer";
import fileInfo from "Files/Redux/FileInfoReducer";
import header, {CONTEXT_SWITCH, USER_LOGIN, USER_LOGOUT} from "Navigation/Redux/HeaderReducer";
import sidebar from "Navigation/Redux/SidebarReducer";
import status from "Navigation/Redux/StatusReducer";
import notifications from "Notifications/Redux/NotificationsReducer";
import * as ProjectRedux from "Project/Redux";
import {Action, AnyAction, combineReducers, createStore, Store} from "redux";
import {composeWithDevTools} from "redux-devtools-extension";
import {createResponsiveStateReducer, responsiveStoreEnhancer} from "redux-responsive";
import simpleSearch from "Search/Redux/SearchReducer";
import {responsiveBP} from "ui-components/theme";
import uploader from "Uploader/Redux/UploaderReducer";
import avatar from "UserSettings/Redux/AvataaarReducer";
import hookStore from "Utilities/ReduxHooks";

export function configureStore(
    initialObject: ReduxObject,
    reducers,
    enhancers?
): Store<ReduxObject, AnyAction> {
    const combinedReducers = combineReducers<ReduxObject, AnyAction>(reducers);
    const rootReducer = (state: ReduxObject, action: Action): ReduxObject => {
        if ([USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH].some(it => it === action.type)) {
            state = initObject();
        }
        return combinedReducers(state, action);
    };
    return createStore<ReduxObject, AnyAction, {}, {}>(rootReducer, initialObject, composeWithDevTools(enhancers));
}

export const responsive = createResponsiveStateReducer(
    responsiveBP,
    {infinity: "xxl"}
);

export const store = configureStore(initObject(), {
    activity,
    dashboard,
    analyses,
    applications,
    header,
    status,
    sidebar,
    uploader,
    notifications,
    simpleSearch,
    detailedFileSearch,
    detailedApplicationSearch,
    fileInfo,
    hookStore,
    ...AppRedux.reducers,
    avatar,
    loading,
    tasks: TaskRedux.reducer,
    project: ProjectRedux.reducer,
    responsive: createResponsiveStateReducer(
        responsiveBP,
        {infinity: "xxl"}),
}, responsiveStoreEnhancer);

function loading(state = false, action: {type: string}): boolean {
    switch (action.type) {
        case "LOADING_START":
            return true;
        case "LOADING_END":
            return false;
        default:
            return state;
    }
}
