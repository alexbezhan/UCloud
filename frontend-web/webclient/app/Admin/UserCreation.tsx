import * as React from "react";
import {Cloud} from "Authentication/SDUCloudObject";
import PromiseKeeper from "PromiseKeeper";
import {defaultErrorHandler} from "UtilityFunctions";
import {UserCreationState} from ".";
import {Input, Label, Button} from "ui-components";
import * as Heading from "ui-components/Heading";
import {connect} from "react-redux";
import {Dispatch} from "redux";
import {setActivePage, SetStatusLoading, setLoading} from "Navigation/Redux/StatusActions";
import {SidebarPages} from "ui-components/Sidebar";
import {MainContainer} from "MainContainer/MainContainer";
import {SnackType} from "Snackbar/Snackbars";
import {snackbarStore} from "Snackbar/SnackbarStore";

const initialState: UserCreationState = {
    submitted: false,
    username: "",
    password: "",
    repeatedPassword: "",
    usernameError: false,
    passwordError: false
};

function UserCreation(props: UserCreationOperations) {
    // Use reducer instead, or break into smaller ones.
    const [state, setState] = React.useState(initialState);
    const [promiseKeeper] = React.useState(new PromiseKeeper());

    React.useEffect(() => {
        props.setActivePage();
        return () => promiseKeeper.cancelPromises();
    }, []);

    function updateFields(field: keyof UserCreationState, value: string) {
        if (field === "username") state.usernameError = false;
        else if (field === "password" || field === "repeatedPassword") state.passwordError = false;
        setState({...state, [field]: value});
    }

    async function submit(e: React.SyntheticEvent) {
        e.preventDefault();

        let usernameError = false;
        let passwordError = false;
        const {username, password, repeatedPassword} = state;
        if (!username) usernameError = true;
        if (!password || password !== repeatedPassword) passwordError = true;
        setState({...state, usernameError, passwordError});
        if (!usernameError && !passwordError) {
            try {
                props.setLoading(true);
                await promiseKeeper.makeCancelable(Cloud.post("/auth/users/register", {username, password}, "")).promise;
                snackbarStore.addSnack({message: `User '${username}' successfully created`, type: SnackType.Success});
                setState(() => initialState);
            } catch (e) {
                const status = defaultErrorHandler(e);
                if (status == 400) {
                    snackbarStore.addSnack({
                        message: "User already exists",
                        type: SnackType.Information
                    });
                    setState({...state, usernameError: true});
                }
            } finally {
                props.setLoading(false);
            }
        }
    }


    if (!Cloud.userIsAdmin) return null;

    const {
        usernameError,
        passwordError,
        username,
        password,
        repeatedPassword,
        submitted
    } = state;


    const header = (<><Heading.h1>User Creation</Heading.h1>
        <p>Admins can create new users on this page.</p></>);

    return (
        <MainContainer
            header={header}
            headerSize={120}
            main={<form onSubmit={e => submit(e)}>
                <Label mb="1em">
                    Username
                    <Input
                        value={username}
                        color={usernameError ? "red" : "gray"}
                        onChange={({target: {value}}) => updateFields("username", value)}
                        placeholder="Username..."
                    />
                </Label>
                <Label mb="1em">
                    Password
                    <Input
                        value={password}
                        type="password"
                        color={passwordError ? "red" : "gray"}
                        onChange={({target: {value}}) => updateFields("password", value)}
                        placeholder="Password..."
                    />
                </Label>
                <Label mb="1em">
                    Repeat password
                    <Input
                        value={repeatedPassword}
                        type="password"
                        color={passwordError ? "red" : "gray"}
                        onChange={({target: {value}}) => updateFields("repeatedPassword", value)}
                        placeholder="Repeat password..."
                    />
                </Label>
                <Button
                    type="submit"
                    color="green"
                    disabled={submitted}
                >Create user</Button>
            </form>}
        />
    );
}

interface UserCreationOperations extends SetStatusLoading {
    setActivePage: () => void
}

const mapDispatchToProps = (dispatch: Dispatch): UserCreationOperations => ({
    setActivePage: () => dispatch(setActivePage(SidebarPages.Admin)),
    setLoading: loading => dispatch(setLoading(loading))
});

export default connect(null, mapDispatchToProps)(UserCreation);
