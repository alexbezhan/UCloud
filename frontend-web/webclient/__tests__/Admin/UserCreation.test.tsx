import {configure, mount} from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import * as React from "react";
import {Provider} from "react-redux";
import {MemoryRouter} from "react-router";
import {create} from "react-test-renderer";
import {ThemeProvider} from "styled-components";
import UserCreation from "../../app/Admin/UserCreation";
import PromiseKeeper from "../../app/PromiseKeeper";
import theme from "../../app/ui-components/theme";
import {store} from "../../app/Utilities/ReduxUtilities";

configure({adapter: new Adapter()});

const userCreation = () => (
    <Provider store={store}>
        <MemoryRouter>
            <ThemeProvider theme={theme}>
                <UserCreation />
            </ThemeProvider>
        </MemoryRouter>
    </Provider>
);

describe("UserCreation", () => {
    test("Mount", () => expect(create(userCreation()).toJSON()).toMatchSnapshot());

    test.skip("Update username field", () => {
        const uC = mount(userCreation());
        uC.find("FormField").findWhere(it => it.props().label === "Username").find("input").simulate("change", {target: {value: "username"}});
        expect(uC.state()).toEqual({
            promiseKeeper: new PromiseKeeper(),
            submitted: false,
            username: "username",
            password: "",
            repeatedPassword: "",
            usernameError: false,
            passwordError: false
        });
    });

    test.skip("Update password field", () => {
        const uC = mount(userCreation());
        uC.find("FormField").findWhere(it => it.props().label === "Password").find("input").simulate("change", {target: {value: "password"}});
        expect(uC.state()).toEqual({
            promiseKeeper: new PromiseKeeper(),
            submitted: false,
            username: "",
            password: "password",
            repeatedPassword: "",
            usernameError: false,
            passwordError: false
        });
    });

    test.skip("Update repeated password field", () => {
        const uC = mount(userCreation());
        uC.find("FormField").findWhere(it => it.props().label === "Repeat password").find("input").simulate("change", {target: {value: "repeatWord"}});
        expect(uC.state()).toEqual({
            promiseKeeper: new PromiseKeeper(),
            submitted: false,
            username: "",
            password: "",
            repeatedPassword: "repeatWord",
            usernameError: false,
            passwordError: false
        });
    });

    test.skip("Submit with missing username, causing errors to be rendered", () => {
        const uC = mount(userCreation());
        uC.find("FormField").findWhere(it => it.props().label === "Password").find("input").simulate("change", {target: {value: "password"}});
        uC.find("FormField").findWhere(it => it.props().label === "Repeat password").find("input").simulate("change", {target: {value: "password"}});
        uC.find("Button").findWhere(it => it.props().content === "Create user").simulate("submit");
        expect(uC.state("usernameError")).toBe(true);
        expect(uC.state("passwordError")).toBe(false);
    });

    test.skip("Submit with missing password fields, causing errors to be rendered", () => {
        const uC = mount(userCreation());
        uC.find("FormField").findWhere(it => it.props().label === "Username").find("input").simulate("change", {target: {value: "username"}});
        uC.find("Button").findWhere(it => it.props().content === "Create user").simulate("submit");
        expect(uC.state("passwordError")).toBe(true);
        expect(uC.state("usernameError")).toBe(false);
    });

    test.skip("Submit with non matching password fields, causing errors to be rendered", () => {
        const uC = mount(userCreation());
        uC.find("FormField").findWhere(it => it.props().label === "Password").find("input").simulate("change", {target: {value: "passwordAlso"}});
        uC.find("FormField").findWhere(it => it.props().label === "Repeat password").find("input").simulate("change", {target: {value: "password"}});
        uC.find("Button").findWhere(it => it.props().content === "Create user").simulate("submit");
        expect(uC.state("usernameError")).toBe(true);
        expect(uC.state("passwordError")).toBe(true);
    });
});
