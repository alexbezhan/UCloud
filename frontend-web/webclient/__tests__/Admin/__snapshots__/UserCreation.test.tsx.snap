// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCreation Mount 1`] = `
.c0 {
  margin-left: 190px;
  padding-top: 48px;
  padding-bottom: 14px;
  padding-left: 14px;
  padding-right: 0;
  background-color: #fff;
}

.c3 {
  padding-top: 64px;
  padding-right: 14px;
}

.c7 {
  font-smoothing: antialiased;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 700;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 5px;
  background-color: var(--green,#f00);
  color: var(--white,#f00);
  border-width: 0;
  border-style: solid;
  -webkit-transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  font-size: 14px;
  padding: 9.5px 18px;
}

.c7:disabled {
  opacity: 0.25;
}

.c7:focus {
  outline: none;
}

.c7:hover {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c2 {
  cursor: inherit;
  font-size: 40px;
  margin: 0;
  font-weight: 700;
}

.c6 {
  display: block;
  font-family: inherit;
  background-color: transparent;
  color: var(--black,#f00);
  margin: 0;
  border-width: 2px;
  border-color: #c9d3df;
  border-style: solid;
  padding-top: 7px;
  padding-bottom: 7px;
  padding-left: 12px;
  padding-right: 12px;
  border-radius: 5px;
  width: 100%;
}

.c6::-webkit-input-placeholder {
  color: var(--gray,#f00);
}

.c6::-moz-placeholder {
  color: var(--gray,#f00);
}

.c6:-ms-input-placeholder {
  color: var(--gray,#f00);
}

.c6::placeholder {
  color: var(--gray,#f00);
}

.c6:focus {
  outline: none;
  background-color: transparent;
}

.c6:focus {
  outline: 0;
  border-color: #006aff;
}

.c5 {
  font-size: 10px;
  -webkit-letter-spacing: 0.2px;
  -moz-letter-spacing: 0.2px;
  -ms-letter-spacing: 0.2px;
  letter-spacing: 0.2px;
  display: block;
  margin: 0;
  margin-bottom: 1em;
  font-size: 14px;
  color: #000;
  font-weight: 700;
  width: 100%;
}

.c1 {
  padding-top: 14px;
  padding-bottom: 14px;
  padding-left: 204px;
  padding-right: 14px;
  width: 100%;
  height: 64px;
  background-color: #fff;
  position: absolute;
  top: 48px;
  left: 0;
  position: fixed;
  z-index: 50;
}

@media screen and (min-width:768px) and (max-width:1023px) {
  .c4 {
    display: none;
  }
}

@media screen and (min-width:1024px) and (max-width:1279px) {
  .c4 {
    display: none;
  }
}

@media screen and (min-width:1280px) {
  .c4 {
    display: none;
  }
}

<div
  className="c0"
>
  <div
    className="sc-AxgMl c1"
    height={64}
    width={1}
  >
    <h1
      className="c2"
      cursor="inherit"
      fontSize={6}
    >
      User Creation
    </h1>
  </div>
  <div
    className="c3"
  >
    <div
      className="c4"
    />
    <p>
      Admins can create new users on this page.
    </p>
    <form
      autoComplete="off"
      onSubmit={[Function]}
    >
      <label
        className="c5"
        color="black"
        fontSize={1}
        fontWeight="bold"
        width="100%"
      >
        Username
        <input
          className="c6"
          id="default"
          onChange={[Function]}
          placeholder="Username..."
          value=""
          width="100%"
        />
      </label>
      <label
        className="c5"
        color="black"
        fontSize={1}
        fontWeight="bold"
        width="100%"
      >
        Password
        <input
          className="c6"
          id="default"
          onChange={[Function]}
          placeholder="Password..."
          type="password"
          value=""
          width="100%"
        />
      </label>
      <label
        className="c5"
        color="black"
        fontSize={1}
        fontWeight="bold"
        width="100%"
      >
        Repeat password
        <input
          className="c6"
          id="default"
          onChange={[Function]}
          placeholder="Repeat password..."
          type="password"
          value=""
          width="100%"
        />
      </label>
      <label
        className="c5"
        color="black"
        fontSize={1}
        fontWeight="bold"
        width="100%"
      >
        Email
        <input
          className="c6"
          id="default"
          onChange={[Function]}
          placeholder="Email..."
          type="email"
          value=""
          width="100%"
        />
      </label>
      <button
        className="c7"
        color="green"
        disabled={false}
        type="submit"
      >
        Create user
      </button>
    </form>
  </div>
</div>
`;
